### ISSUE TITLE
Bound type variables are conflated with literal types

### ISSUE THREAD

**Issue by fedorg** [microsoft/pyright#217]
**Describe the bug**
False positive squigglies on what I think is a correctly typed usage of bound TypeVar.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/4352852/63506375-0d834000-c4de-11e9-8e12-ec0652e0983b.png)

```python
from typing import TypeVar, Generic, Type, Union


BaseT = TypeVar('BaseT', bound='Base')

class Base:
    basething: str = 'i am basething'
    _basesec: str = 'basesec'

    def __init__(self):
        self.basething = 'asdfasdfa'

    def dostuff(self, other: Type[BaseT], oi: BaseT):
        print(other.basething)
        print(oi.basething)
        self.basething = 'changed_basething'


b = Base()
o = Base()

b.dostuff(Base, o)
print(b.basething, o.basething, Base.basething)

```

**VS Code extension or command-line**
Pyright 1.0.56 Extension (no CLI)
VSCode 1.38.0-insider
I'm using the `mypy` linter and these workspace settings:
```
    "python.jediEnabled": false,
    "python.linting.enabled": true,
    "python.linting.mypyEnabled": true,
    "python.linting.pylintEnabled": false
```

**Comment by erictraut** [comment]
Hmm, this is debatable. Pyright is treating `'Base'` as a string literal rather than a forward reference to a type. That's because it's being used in a context (an argument value in a function call) where it's perfectly legitimate to pass string literal values. In places within the syntax where a type annotation is expected (and _only_ a type annotation can be specified), type checkers are supposed to assume that a string literal should be treated as a forward reference type expression. Argument values do not meet this requirement, so pyright is treating as just a string literal.

One could argue that in this very specific case of the "bound" parameter for the "TypeVar" constructor, the type checker could assume that the argument should be interpreted as a type expression even if it receives a string literal, but that's a pretty ugly special case to add to the parsing logic. The parsing logic shouldn't know anything about specific calls or arguments.

(As an aside, the whole forward reference mechanism in Python is really screwed up. Python 3.8 is looking to fix it by not evaluating type annotations. This would eliminate the need to specify forward-reference type annotations using string literals. But you still won't be able to do what you're trying to do above because arguments to calls will still be evaluated by the Python 3.8 interpreter, and forward references will raise exceptions.)

In your particular example above, I think there's a much easier solution than what you're trying to do with the TypeVar. Does this accomplish what you're trying to do?
```
import collections

Person = collections.namedtuple("Person", ["name", "age", "gender"])
bily = Person(name="Bily", age=28, gender="male")

class Base:
    basething: str = 'i am basething'
    _basesec: str = 'basesec'

    def __init__(self):
        self.basething = 'asdfasdfa'

    def dostuff(self, oi: 'Base'):
        print(oi.__class__.basething)
        print(oi.basething)
        self.basething = 'changed_basething'

b = Base()
o = Base()

b.dostuff(o)
print(b.basething, o.basething, Base.basething)
```

**Comment by fedorg** [comment]
Thank you for such a detailed explanation. I can see how TypeVar is such an odd type constructor.
However, what I'm trying to do is a little bit less trivial and goes to show how silly of an example I've picked. I'm certainly putting too much stake (and then some) in Python's bolt-on, ad-hoc type system :)

What I have is an ORM-like model around a REST service (I know about OpenAPI but it's not a great fit).
My code has a `Base` class from which all entities inherit.
I want to be able to express features like deserialization `(Type[BaseT], JSON) -> List[BaseT]`, references `Reference[Type[BaseT]] -> Optional[BaseT]`, filter, search, etc., that the `Base` type provides for its descendants.
References to own type naturally require forward annotations, so there is `from __future__ import annotations` everywhere.

So here is a more elaborate example of what the code looks like (yes, the real code does work, and it even looks pretty):
```python
# orm.py
from __future__ import annotations
from typing import TypeVar, Generic, Type, List, Dict, Optional, Any
from dataclasses import dataclass, fields

BaseT = TypeVar('BaseT', bound='Base')

class Reference(Generic[BaseT]):
    """A notion of a GUID reference to a remote object instance"""
    _cls: Type[BaseT]
    _id: str
    
    def __init__(self: Reference[BaseT], cls_: Type[BaseT], id_: str):
        self._cls = cls_
        self._id = id_

    def get(self: Reference[BaseT]) -> List[BaseT]:
        stub = {f.name: '' for f in fields(self._cls)} # ...fetch, deserialize, resolve refs, etc ...
        stub["id"] = self._id
        return [self._cls(**stub)] 


@dataclass
class Base:
    id: str

    def ref(self: BaseT) -> Reference[BaseT]:
        return Reference(type(self), self.id)

    @classmethod
    def from_dict(cls: Type[BaseT], obj: Dict[str, Any], throw=False) -> BaseT:
        return cls(**obj) # ...not really that simple :)

# usage.py

@dataclass
class Group(Base):
    users: Optional[Reference[User]] = None


@dataclass
class User(Base):
    login: str
    manager: Reference[User]
    group: Reference[Group]

g = Group(id='id_group0', users=None)
u = User(id='id_user1', login='user1', manager=Reference(User, 'id_user2'), group=g.ref())
manager = u.manager.get()[0]
print(manager.id)
```

**Comment by erictraut** [comment]
Thanks for the additional details.

You're using an anti-pattern here. You have two classes (`Reference` and `Base`) that have deep knowledge of each other, creating a circular dependency.

Here's a suggested approach that eliminates the circularity. With this change, the `Base` class no longer has any knowledge of `Reference`.
```
@dataclass
class Base:
    id: str

    @classmethod
    def from_dict(cls: Type[BaseT], obj: Dict[str, Any], throw=False) -> BaseT:
        return cls(**obj) # ...not really that simple :)

BaseT = TypeVar('BaseT', bound=Base)


class Reference(Generic[BaseT]):
    """A notion of a GUID reference to a remote object instance"""
    _cls: Type[BaseT]
    _id: str
    
    def __init__(self, cls_: Type[BaseT], id_: str):
        self._cls = cls_
        self._id = id_

    def get(self) -> List[BaseT]:
        stub = {f.name: '' for f in fields(self._cls)} # ...fetch, deserialize, resolve refs, etc ...
        stub["id"] = self._id
        return [self._cls(**stub)] 

    @staticmethod
    def make_ref(val: BaseT) -> Reference[BaseT]:
        return Reference(val.__class__, val.id)
```

(Note that I've also eliminated the type annotations for the `self` parameters. These are not needed because they can always be inferred by type checkers like mypy and pyright.)

**Comment by BjMrq** [comment]
hello, I might have a somewhat simple question, but I have not to find the answer, in 
```def from_dict(cls: Type[BaseT], obj: Dict[str, Any], throw=False) -> BaseT:
        return cls(**obj) # ...not really that simple :)```

how can you use BaseT before it is defined?

**Comment by erictraut** [comment]
Thanks for asking. My sample was not complete as posted.

You have two options:
1. Enclose the type annotations that contain `BaseT` in quotes, which indicates that they include forward references.
2. Add `from __future__ import annotations` at the top of the file, which enables forward references for all type annotations.

**Comment by BjMrq** [comment]
Hi @erictraut thanks a lot for your quick answer!
